mpcs/src/basefold/query_phase.rs

```diff
@@ -61,173 +64,408 @@ where
+                fixed_opened_values
+            } else {
+                &vec![]
+            };
+
+            let mut fixed_commit_leafs_iter = fixed_commit_leafs.iter();
+            let mut batch_coeffs_iter = batch_coeffs.iter();
+
+            let base_codeword_lo_hi = circuit_meta
+                .iter()
+                .zip_eq(witin_opened_values)
+                .map(
+                    |(
+                        CircuitIndexMeta {
+                            witin_num_polys,
+                            fixed_num_vars,
+                            fixed_num_polys,
+                            ..
+                        },
+                        witin_leafs,
+                    )| {
+                        let (lo, hi) = std::iter::once((witin_leafs, *witin_num_polys))
+                            .chain((*fixed_num_vars > 0).then(|| {
+                                (fixed_commit_leafs_iter.next().unwrap(), *fixed_num_polys)
+                            }))
+                            .map(|(leafs, num_polys)| {
+                                let batch_coeffs = batch_coeffs_iter
+                                    .by_ref()
+                                    .take(num_polys)
+                                    .copied()
+                                    .collect_vec();
+                                let (lo, hi): (&[E::BaseField], &[E::BaseField]) =
+                                    leafs.split_at(leafs.len() / 2);
+                                (
+                                    dot_product::<E, _, _>(
+                                        batch_coeffs.iter().copied(),
+                                        lo.iter().copied(),
+                                    ),
+                                    dot_product::<E, _, _>(
+                                        batch_coeffs.iter().copied(),
+                                        hi.iter().copied(),
+                                    ),
+                                )
+                            })
+                            // fold witin/fixed lo, hi together because they share the same num_vars
+                            .reduce(|(lo_wit, hi_wit), (lo_fixed, hi_fixed)| {
+                                (lo_wit + lo_fixed, hi_wit + hi_fixed)
+                            })
+                            .expect("unreachable");
+                        (lo, hi)
+                    },
+                )
+                .collect_vec();
+            debug_assert_eq!(folding_sorted_order.len(), base_codeword_lo_hi.len());
+            debug_assert!(fixed_commit_leafs_iter.next().is_none());
+            debug_assert!(batch_coeffs_iter.next().is_none());
+
+            // fold and query
+            let mut cur_num_var = max_num_var;
+            let rounds = cur_num_var
+                - <Spec::EncodingScheme as EncodingScheme<E>>::get_basecode_msg_size_log()
+                - 1;
```

Comment:
why do we need to minus 1?

Reply:
add explanation in https://github.com/scroll-tech/ceno/pull/894/commits/3c20a048416b5965f92a4ff88a66ebeca79fbe3e